project/UI/CLIKit/ParsableCommand+IO.swift

```diff
@@ -0,0 +1,54 @@
+//  Copyright © 2020 The nef Authors.
+
+import Foundation
+import ArgumentParser
+import Bow
+import BowEffects
+
+public protocol ConsoleCommand: ParsableCommand {
+    func main() -> IO<CLIKit.Console.Error, Void>
+}
+
+extension ParsableCommand {
+    
+    static func parseArguments() -> IO<Console.Error, ParsableCommand> {
+        parseIO()
+            .flatMap { command in command.runIO() }^
+            .mapError(fullMessage)^
+    }
+    
+    static func fullMessage(_ e: Swift.Error) -> Console.Error {
+        .arguments(info: fullMessage(for: e))
+    }
+    
+    static func helpError() -> Swift.Error {
+        CleanExit.helpRequest(self)
+    }
+    
+    // MARK: helpers
+    private static func parseIO() -> IO<Swift.Error, ParsableCommand> {
+        IO.invoke { try parseAsRoot() }
+    }
+    
+    private func runIO() -> IO<Swift.Error, ParsableCommand> {
+        IO.invoke { try self.run() }.map { _ in self }^
```

Comment:
I don't understand why you are running the command and then mapping it back to self.

Reply:
because the `run` (ParsableCommand) signature is:
`func run() throws`
so `IO.invoke` type is - `IO<Swift.Error, Void>` and I need to be `ParsableCommand` instead to `Void`

Reply:
the are mutating the same ParsableCommand with the parsed arguments... 😓

Reply:
It is the place where we swizzling the default `run` behaviour, the run method does not do anything but it can fail.

Reply:
Why isn't it then `IO.pure(self)`? If you run it and then remap it, you are allowing it to be run again accidentally, and I guess that is not what you want this method to do.

Reply:
because the success will be ignored, but it can throw errors - it should be the same to:
```swift
IO.invoke { 
   try self.run()
   return self
}
```
maybe it is more clear, I do not know